processing/src/main/java/org/apache/druid/frame/processor/OutputChannel.java

```diff
@@ -0,0 +1,162 @@
+   * @param readableChannelSupplier readable channel for consumer. May be called multiple times, so you should wrap this
+   *                                in {@link Suppliers#memoize} if needed.
+   * @param partitionNumber         partition number, if any; may be {@link FrameWithPartition#NO_PARTITION} if unknown
+   */
+  public static OutputChannel pair(
+      final WritableFrameChannel writableChannel,
+      final MemoryAllocator frameMemoryAllocator,
+      final Supplier<ReadableFrameChannel> readableChannelSupplier,
+      final int partitionNumber
+  )
+  {
+    return new OutputChannel(
+        Preconditions.checkNotNull(writableChannel, "writableChannel"),
+        Preconditions.checkNotNull(frameMemoryAllocator, "frameMemoryAllocator"),
+        readableChannelSupplier,
+        partitionNumber
+    );
+  }
+
+  /**
+   * Create a nil output channel, representing a processor that writes nothing. It is not actually writable, but
+   * provides a way for downstream processors to read nothing.
+   */
+  public static OutputChannel nil(final int partitionNumber)
+  {
+    return new OutputChannel(null, null, () -> ReadableNilFrameChannel.INSTANCE, partitionNumber);
+  }
+
+  /**
+   * Returns the writable channel of this pair. The producer writes to this channel.
+   */
+  public WritableFrameChannel getWritableChannel()
+  {
+    if (writableChannel == null) {
+      throw new ISE("Writable channel is not available");
+    } else {
+      return writableChannel;
+    }
+  }
+
+  /**
+   * Returns the memory allocator for the writable channel. The producer uses this to generate frames for the channel.
+   */
+  public MemoryAllocator getFrameMemoryAllocator()
+  {
+    if (frameMemoryAllocator == null) {
+      throw new ISE("Writable channel is not available");
+    } else {
+      return frameMemoryAllocator;
+    }
+  }
+
+  /**
+   * Returns the readable channel of this pair. This readable channel may, or may not, be usable before the
+   * writable channel is closed. It depends on whether the channel pair was created in a stream-capable manner or not.
+   */
+  public ReadableFrameChannel getReadableChannel()
+  {
+    return readableChannelSupplier.get();
+  }
+
+  public int getPartitionNumber()
+  {
+    return partitionNumber;
+  }
+
+  public OutputChannel mapWritableChannel(final Function<WritableFrameChannel, WritableFrameChannel> mapFn)
+  {
+    if (writableChannel == null) {
+      return this;
+    } else {
+      return new OutputChannel(
+          mapFn.apply(writableChannel),
+          frameMemoryAllocator,
+          readableChannelSupplier,
+          partitionNumber
+      );
+    }
+  }
+
+  /**
+   * Returns a read-only version of this instance. Read-only versions have neither {@link #getWritableChannel()} nor
+   * {@link #getFrameMemoryAllocator()}, and therefore require substantially less memory.
+   */
+  public OutputChannel readOnly()
```

Comment:
OK, I'll bite. What does "read only output channel" even mean? Doesn't output kinda imply we're doing some writing and not doing any reading?

Reply:
OutputChannel is really a pair of a readable and a writable channel. The idea is an OutputChannel is not just used for the output of some processor, it's also used so some _other_ processor can _read_ the output, so there's the readable channel too.

A read only instance doesn't have the writable channel.

Open to ideas of what a better name would be.

Reply:
I kind of like the Go approach: there is a channel, then a channel can provide a channel reader and a channel writer. In Go, only the writer can close the channel (to indicate EOF), the reader doesn't have to. That seems like a wise choice: there can be multiple readers, and the Go approach avoids the need to track them to detect when the last reader has closed its channel reader.

So, one thought would be to replicate that idea here. A "channel" is the pipe. A "channel reader" is the output and a "channel writer" is the input. A channel with only readers is expressed simply in this model.

This is another of those "make it easy for developers to understand" things, not functional, so it's fine to defer that kind of refactoring to later.

Reply:
Hmm. Maybe this interface should just be called "Channel", or "FrameChannel", then? Since we already have ReadableFrameChannel and WritableFrameChannel for the readable and writable halves.